build.gradle

```diff
@@ -279,6 +295,78 @@ subprojects {
       exceptionFormat = testExceptionFormat
     }
     logTestStdout.rehydrate(delegate, owner, this)()
+
+    afterTest { TestDescriptor desc , TestResult result ->
+      if (TestResult.ResultType.FAILURE == result.resultType) {
+        logger.warn("Found failing test $desc.parent - $desc.name, trying again in next turn of test execution")
+        failedTestsInFirstTestRun << desc
+      }
+    }
+    testResultsDir = reports.junitXml.destination
+
+    ignoreFailures = shouldFilterFlakyTests
+  }
+
+  task testWithFlakyRetry(type: Test, dependsOn: test) {
+    maxParallelForks = userMaxForks ?: Runtime.runtime.availableProcessors()
+
+    minHeapSize = "256m"
+    maxHeapSize = "2048m"
+
+    testLogging {
+      events = userTestLoggingEvents ?: testLoggingEvents
+      showStandardStreams = userShowStandardStreams ?: testShowStandardStreams
+      exceptionFormat = testExceptionFormat
+    }
+
+    def oldResultsToOverwite = []
+
+    doFirst {
+      if (!failedTestsInFirstTestRun.isEmpty()) {
+        logger.warn("Executing failed tests again to filter out flaky tests")
+        logger.warn("Failed tests: " + failedTestsInFirstTestRun)
+
+        logger.warn("Successful results will overwrite stale failed test results from $testResultsDir to avoid jenkins reporting an UNSTABLE build")
+        def failedTestClasses = failedTestsInFirstTestRun.collect{
+          it.className.replaceAll('\\.', '/') + ".*"
+        } as Set
+        // re-running complete classes to keep the integrity of the test results
+        logger.warn("Re-running $failedTestClasses")
```

Comment:
nit: would it be more readable if we showed the amount of failing classes first and then iterated them? something like `Re-running 5 failing test classes (a, b, c, d, e)`

Reply:
Makes sense

Reply:
Also I think it'd be enough to print the number of classes as the failed tests will be printed anyway just above